src/Nulls.jl

```diff
@@ -115,9 +115,170 @@ xor(::Integer, ::Null) = null
+ 1
+ 0
+ 2
+
+julia> collect(Nulls.replace([1 null; 2 null], 0))
+2×2 Array{Int64,2}:
+ 1  0
+ 2  0
+
+```
+"""
+replace(itr, replacement) = EachReplaceNull(itr, replacement)
+struct EachReplaceNull{T, U}
+    x::T
+    replacement::U
+end
+Base.iteratorsize(::Type{<:EachReplaceNull{T}}) where {T} =
+    Base.iteratorsize(T)
+Base.iteratoreltype(::Type{<:EachReplaceNull{T}}) where {T} =
+    Base.iteratoreltype(T)
+Base.length(itr::EachReplaceNull) = length(itr.x)
+Base.size(itr::EachReplaceNull) = size(itr.x)
+Base.start(itr::EachReplaceNull) = start(itr.x)
+Base.done(itr::EachReplaceNull, state) = done(itr.x, state)
+Base.eltype(itr::EachReplaceNull) =
+    Union{Nulls.T(eltype(itr.x)), typeof(itr.replacement)}
+@inline function Base.next(itr::EachReplaceNull, state)
+    v, s = next(itr.x, state)
+    ((isnull(v) ? itr.replacement : v)::eltype(itr), s)
+end
+
+"""
+    Nulls.skip(itr)
+
+Return an iterator wrapping iterable `itr` which skips [`null`](@ref) values.
+
+Use [`collect`](@ref) to obtain an `Array` containing the non-`null` values in
+`itr`. Note that even if `itr` is a multidimensional array, the result will always
+be a `Vector` since it is not possible to remove nulls while preserving dimensions
+of the input.
+
+See also: [`Nulls.replace`](@ref), [`Nulls.fail`](@ref)
+
+# Examples
+```jldoctest
+julia> collect(Nulls.skip([1, null, 2]))
+2-element Array{Int64,1}:
+ 1
+ 2
+
+julia> collect(Nulls.skip([1 null; 2 null]))
+2-element Array{Int64,1}:
+ 1
+ 2
+
+```
+"""
+skip(itr) = EachSkipNull(itr)
+struct EachSkipNull{T}
+    x::T
+end
+Base.iteratorsize(::Type{<:EachSkipNull}) =
+    Base.SizeUnknown()
+Base.iteratoreltype(::Type{EachSkipNull{T}}) where {T} =
+    Base.iteratoreltype(T)
+Base.eltype(itr::EachSkipNull) = Nulls.T(eltype(itr.x))
+# Fallback implementation for general iterables: we cannot access a value twice,
+# so after finding the next non-null element in start() or next(), we have to
+# pass it in the iterator state, which introduces a type instability since the value
+# is null if the input does not contain any non-null element.
+# As of Julia 0.6 and early 0.7, this instability kills performance.
+@inline function Base.start(itr::EachSkipNull)
+    s = start(itr.x)
+    v = null
+    @inbounds while !done(itr.x, s) && isnull(v)
+        v, s = next(itr.x, s)
+    end
+    (v, s)
+end
+@inline Base.done(itr::EachSkipNull, state) = isnull(state[1]) && done(itr.x, state[2])
+@inline function Base.next(itr::EachSkipNull, state)
+    v1, s = state
+    v2 = null
+    @inbounds while !done(itr.x, s) && isnull(v2)
+        v2, s = next(itr.x, s)
+    end
+    (v1::eltype(itr), (v2, s))
+end
+# Optimized implementation for AbstractArray, relying on the ability to access x[i] twice:
+# once in done() to find the next non-null entry, and once in next() to return it.
+# This works around the type instability problem of the generic fallback.
+@inline function _next_nonnull_ind(x::AbstractArray, s)
+    idx = eachindex(x)
```

Comment:
I couldn't find a simpler way to work with indices which would be completely generic. Linear indices would of course work, but they would be slow for `LinearSlow` arrays.

Reply:
Couldn't we dispatch and use linear indices for most arrays but a specialized implementation for `LinearSlow`?